File: Mango.Services.AuthAPI/Service/AuthService.cs
using Xango.Services.AuthAPI.Data;
using Xango.Services.AuthAPI.Models;
using Xango.Services.AuthAPI.Service.IService;
using Microsoft.AspNetCore.Identity;
using AutoMapper;
using Xango.Models.Dto;
using Xango.Services.Dto;

namespace Xango.Services.AuthAPI.Service
{
    public class AuthService : IAuthService
    {
        private readonly AppDbContext _db;
        private readonly UserManager<ApplicationUser> _userManager;
        private readonly RoleManager<IdentityRole> _roleManager;
        private readonly IJwtTokenGenerator _jwtTokenGenerator;
        private readonly IHttpContextAccessor _httpContextAccessor;
        private readonly IMapper _mapper;
        protected ResponseDto _response;

        public AuthService(AppDbContext db, IJwtTokenGenerator jwtTokenGenerator,
            UserManager<ApplicationUser> userManager, RoleManager<IdentityRole> roleManager, IHttpContextAccessor ctx, IMapper mapper)
        {
            _db = db;
            _jwtTokenGenerator = jwtTokenGenerator;
            _userManager = userManager;
            _roleManager = roleManager;
            _httpContextAccessor = ctx;
            _response = new();
            _mapper = mapper;
        }

        public async Task<UserDto> GetUser(string email)
        {
            var user = _db.ApplicationUsers.FirstOrDefault(u => u.Email.ToLower() == email.ToLower());
            if (user == null)
            {
                return new UserDto()
                {
                };
            }
            return new UserDto()
            {
                Id = user.Id,
                Email = user.Email,
                Name = user.Name,
                PhoneNumber = user.PhoneNumber
            };
        }

        public async Task<LoginResponseDto> Login(LoginRequestDto loginRequestDto)
        {
            var responseDto = new ResponseDto();
            var user = _db.ApplicationUsers.FirstOrDefault(u => u.UserName.ToLower() == loginRequestDto.UserName.ToLower());

            bool isValid = await _userManager.CheckPasswordAsync(user, loginRequestDto.Password);

            if (user == null || isValid == false)
            {
                return new LoginResponseDto() { User = null, Token = "" };
            }

            //if user was found , Generate JWT Token
            var roles = await _userManager.GetRolesAsync(user);
            var token = _jwtTokenGenerator.GenerateToken(user, roles);
            var userDto = _mapper.Map<UserDto>(user);
            LoginResponseDto loginResponseDto = new LoginResponseDto()
            {
                User = userDto,
                Token = token
            };
            return loginResponseDto;
        }

        public async Task<string?> Register(RegistrationRequestDto registrationRequestDto)
        {
            ApplicationUser user = new()
            {
                UserName = registrationRequestDto.Email,
                Email = registrationRequestDto.Email,
                NormalizedEmail = registrationRequestDto.Email.ToUpper(),
                Name = registrationRequestDto.Name,
                PhoneNumber = registrationRequestDto.PhoneNumber
            };

            try
            {
                var result = await _userManager.CreateAsync(user, registrationRequestDto.Password);
                if (result.Succeeded)
                {
                    var userToReturn = _db.ApplicationUsers.First(u => u.UserName == registrationRequestDto.Email);
                    var userDto = _mapper.Map<UserDto>(userToReturn);

                    return "Registration successful";
                }
                else
                {
                    return result.Errors.FirstOrDefault().Description;
                }

            }
            catch (Exception ex)
            {
                return "Error occurred: " + ex.Message;
            }
        }

        public async Task<bool> AssignRole(string email, string role)
        {
            var user = _db.ApplicationUsers.FirstOrDefault(u => u.Email.ToLower() == email.ToLower());
            if (user != null)
            {
                var roleExists = await _roleManager.RoleExistsAsync(role);
                if (!roleExists)
                {
                    await _roleManager.CreateAsync(new IdentityRole(role));
                }
                await _userManager.AddToRoleAsync(user, role);
                return true;
            }
            return false;
        }
    }
}

File: Xango.React.Application/src/services/authService.ts
import { apiClient, showApiError, unwrapResponse } from './apiClient';
import { serviceSettings } from '../config/serviceSettings';
import type { RegistrationRequest, ResponseDto, UserRecord, UserResponse } from '../types';

export async function login(userName: string, password: string) {
  try {
    const { data } = await apiClient.post<ResponseDto<UserResponse>>(
      `${serviceSettings.AUTH_API}/api/auth/login`,
      { userName, password }
    );
    const result = unwrapResponse(data);
    const token = result.token ?? result.userToken?.token;
    if (!token) {
      throw new Error('Login response did not include a token');
    }
    return { token, user: result.user };
  } catch (error) {
    showApiError(error, 'Login failed');
  }
}

export async function register(registrationRequest: RegistrationRequest) {
  try {
    const { data } = await apiClient.post<ResponseDto>(
      `${serviceSettings.AUTH_API}/api/auth/register`,
      registrationRequest
    );
    return unwrapResponse(data);
  } catch (error) {
    showApiError(error, 'Registration failed');
  }
}

export async function isAdmin(user: UserRecord | null) {
  if (!user) {
    return false;
  }
  try {
    const { data } = await apiClient.get<ResponseDto<boolean>>(
      `${serviceSettings.AUTH_API}/api/auth/InRole/${user.email}/ADMIN`
    );
    return unwrapResponse(data);
  } catch {
    return false;
  }
}
